src/uct/ib/base/ib_device.c

```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
     }
 }
 
-uct_ib_address_scope_t uct_ib_address_scope(uint64_t subnet_prefix)
+uct_ib_address_type_t uct_ib_address_scope(uint64_t subnet_prefix)
 {
     if (subnet_prefix == UCT_IB_LINK_LOCAL_PREFIX) {
-        return UCT_IB_ADDRESS_SCOPE_LINK_LOCAL;
+        return UCT_IB_ADDRESS_TYPE_LINK_LOCAL;
     } else if ((subnet_prefix & UCT_IB_SITE_LOCAL_MASK) == UCT_IB_SITE_LOCAL_PREFIX) {
-        return UCT_IB_ADDRESS_SCOPE_SITE_LOCAL;
+        return UCT_IB_ADDRESS_TYPE_SITE_LOCAL;
     } else {
-        return UCT_IB_ADDRESS_SCOPE_GLOBAL;
+        return UCT_IB_ADDRESS_TYPE_GLOBAL;
     }
 }
 
-size_t uct_ib_address_size(uct_ib_address_scope_t scope)
+size_t uct_ib_address_size(uct_ib_address_type_t type)
 {
-    switch (scope) {
-    case UCT_IB_ADDRESS_SCOPE_LINK_LOCAL:
+    switch (type) {
+    case UCT_IB_ADDRESS_TYPE_LINK_LOCAL:
         return sizeof(uct_ib_address_t) +
                sizeof(uint16_t); /* lid */
-    case UCT_IB_ADDRESS_SCOPE_SITE_LOCAL:
+    case UCT_IB_ADDRESS_TYPE_SITE_LOCAL:
         return sizeof(uct_ib_address_t) +
                sizeof(uint16_t) + /* lid */
                sizeof(uint64_t) + /* if_id */
                sizeof(uint16_t);  /* subnet16 */
-    case UCT_IB_ADDRESS_SCOPE_GLOBAL:
+    case UCT_IB_ADDRESS_TYPE_GLOBAL:
         return sizeof(uct_ib_address_t) +
                sizeof(uint16_t) + /* lid */
                sizeof(uint64_t) + /* if_id */
-               sizeof(uint64_t);  /* subney64 */
+               sizeof(uint64_t);  /* subnet64 */
+    case UCT_IB_ADDRESS_TYPE_ETH_LOCAL:
+        return sizeof(uct_ib_address_t) +
+               16 * sizeof(uint8_t);  /* raw gid */
     default:
-        ucs_fatal("Invalid IB address scope: %d", scope);
+        ucs_fatal("Invalid IB address type: %d", type);
     }
 }
 
-void uct_ib_address_pack(uct_ib_device_t *dev, uct_ib_address_scope_t scope,
+void uct_ib_address_pack(uct_ib_device_t *dev, uct_ib_address_type_t type,
                          const union ibv_gid *gid, uint16_t lid,
                          uct_ib_address_t *ib_addr)
 {
     void *ptr = ib_addr + 1;
 
     ib_addr->flags  = 0;
 
-    /* LID */
-    /* TODO check IB link layer of the port */
-    ib_addr->flags |= UCT_IB_ADDRESS_FLAG_LID;
-    *(uint16_t*)ptr = lid;
-    ptr += sizeof(uint16_t);
-
-    if (scope >= UCT_IB_ADDRESS_SCOPE_SITE_LOCAL) {
-        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_IF_ID;
-        *(uint64_t*)ptr = gid->global.interface_id;
-        ptr += sizeof(uint64_t);
-
-        if (scope >= UCT_IB_ADDRESS_SCOPE_GLOBAL) {
-            /* Global */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET64;
-            *(uint64_t*)ptr = gid->global.subnet_prefix;
-        } else {
-            /* Site-local */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
-            *(uint16_t*)ptr = gid->global.subnet_prefix >> 48;
+    if (type < UCT_IB_ADDRESS_TYPE_ETH_LOCAL) {
```

Comment:
let's not assume order of enum here.
if (type == ETH) {
  // eth code
} else {
  /* IB */
}